### ISSUE TITLE
/usr/bin/ld: ./.libs/libtpms.so: ABI version 2 is not compatible with ABI version 0 output

### ISSUE THREAD

**Issue by frediz** [stefanberger/libtpms#298]
Hi,
this message occurs on Debian (Bullseye/11) ppc64el when building latest upstream ( 967e52e ).

```
make  check-local
make[3]: Entering directory '/<<PKGBUILDDIR>>/src'
There are undefined symbols in libtpms (    )
/usr/bin/ld: ./.libs/libtpms.so: ABI version 2 is not compatible with ABI version 0 output
/usr/bin/ld: failed to merge target specific data of file ./.libs/libtpms.so
make[3]: *** [Makefile:3663: check-local] Error 1
```
**Versions of relevant components**
- libtpms: 0.10.0~dev1 but also with 0.9.0
- openssl: 1.1.1k-1+deb11u1
- gcc version 10.2.1 20210110 (Debian 10.2.1-6)

F.

**Comment by stefanberger** [comment]
You have to show me the exact configure line you have used to build it.

**Comment by stefanberger** [comment]
Also please attach the full build log.

**Comment by frediz** [comment]
Hi Stefan,
the one generated using debian packaging's build, that is :

```
./configure --build=powerpc64le-linux-gnu --prefix=/usr --includedir=\${prefix}/include --mandir=\${prefix}/share/man --infodir=\${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --disable-option-checking --disable-silent-rules --libdir=\${prefix}/lib/powerpc64le-linux-gnu --libexecdir=\${prefix}/lib/powerpc64le-linux-gnu --disable-maintainer-mode --disable-dependency-tracking --with-openssl --with-tpm2
```

**Comment by stefanberger** [comment]
I have build packages for ppc64 for a long time in my Luanchpad PPA: https://launchpad.net/~stefanberger

Only recently with gcc-11 or gcc-12, not sure which one it is using actually, did compiler issues appear: #297  [ https://launchpad.net/~stefanberger/+archive/ubuntu/swtpm-jammy/+build/23174046 ]

So I am bit surprised about the build failure at the linking stage even. But I would still like to see the whole build log.

**Comment by stefanberger** [comment]
Also, do you build it on other architectures as well and they all finish without errors ?

**Comment by frediz** [comment]
Here is the build log on Unstable with 0.9 : 
https://buildd.debian.org/status/fetch.php?pkg=libtpms&arch=ppc64el&ver=0.9.2-2&stamp=1645522481&raw=0

Yes it seems to build on all other supported architectures (non greyed in the following page) :  
https://buildd.debian.org/status/package.php?p=libtpms&suite=sid

I tried myself on amd64 and it built well.

Now on the Ubuntu side, gcc used is gcc-9 and 2.34-6ubuntu1 for focal.
On Jammy where the toolchain should be much newer, we see it fails before.. so maybe it fails actually as on Debian with a toolchain of a closer level to Jammy (Debian Bullseye/11 uses gcc 10 and 2.35.2-2).

**Comment by stefanberger** [comment]
Here's the build on Fedora with powerpc64. It was using gcc-11: https://koji.fedoraproject.org/koji/buildinfo?buildID=1875007

gcc-11.2.1 was installed: https://kojipkgs.fedoraproject.org//packages/libtpms/0.9.2/0.20220106gite81d634c27.fc35.0/data/logs/ppc64le/root.log


I tried this now in an environment similar to yours and I do see the same error with gcc. It seems to be a gcc issue.

My suggestion would to either:
- Use clang if you can. This works as expected: 
```
CC=clang ./autogen.sh --prefix=/usr --with-openssl --with-tpm2
make -j32
make -j32 check
```

- Apply this patch here as work-around for the moment :

```
diff --git a/src/Makefile.am b/src/Makefile.am
index c2759a8..b2e2cdc 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -599,7 +599,7 @@ LDFLAGS_ARCH += $(findstring -m32, $(AM_LDFLAGS))
 LDFLAGS_ARCH += $(findstring -m64, $(AM_LDFLAGS))

 check-local: SHELL?="/usr/bin/env bash"
-check-local:
+NO-check-local:
        @case $(host_os) in \
          openbsd*) ADDLIBS="-lc" ;; \
          darwin*|freebsd*) LDFLAGS_OS="-shared" ;; \
```

- do not run `make check` on powerpc64 platform for now
- leave it broken on powerpc64 and wait until gcc comes to its senses

**Comment by stefanberger** [comment]
I am fixing the `-fstack-protector` issue that some platforms seem to have an issue with. That fix should be in 0.9.3 then. #299

**Comment by stefanberger** [comment]
@frediz  I will release libtpms v0.9.3 today. I already built it on my Ubuntu PPA with two patches applied due to 2 bugs in ppc64 gcc on Ubuntu Jammy only. One has to do with inlining of a function `MakeIv` (#297) and the other one is actually the same as you reported here.

The packages are here: https://launchpad.net/~stefanberger/+archive/ubuntu/swtpm-jammy/+packages

https://launchpad.net/~stefanberger/+archive/ubuntu/swtpm-jammy/+sourcefiles/libtpms/0.9.3-1/libtpms_0.9.3-1.tar.xz

The two patches are already applied using dquilt and can be removed using `dquilt pop -a`.
You should be able to use the tar.gz package or at least extract the `debian/patches/` directory from it with these two patches.

[ I setup dquilt using section 3.1 here:  https://www.debian.org/doc/manuals/maint-guide/modify.en.html ]

In the medium/long-term those two patches will hopefully disappear when the ppc64 gcc bugs gets fixed.

**Comment by frediz** [comment]
Hi @stefanberger , thanks!
I confirm it does build now with your patches on Debian Unstable.
I'll forward this to the maintainer in Debian.
F.

**Comment by stefanberger** [comment]
For the record. This patch here resolves this issue [I will not apply it upstream]:

```
Do not run check-local due to bug in ppc64 gcc on Jammy

ppc64 gcc has a bug causing issues in check-local.
Therefore, do not run check-local while ppc64 gcc is broken
on Ubuntu Jammy.

---
 src/Makefile.am |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -599,7 +599,7 @@
 LDFLAGS_ARCH += $(findstring -m64, $(AM_LDFLAGS))

 check-local: SHELL?="/usr/bin/env bash"
-check-local:
+NO-check-local:
        @case $(host_os) in \
          openbsd*) ADDLIBS="-lc" ;; \
          darwin*|freebsd*) LDFLAGS_OS="-shared" ;; \
```

**Comment by stefanberger** [comment]
I released v0.9.3 yesterday. It is basically the same as the package I had already build for my PPA and pointed to above:

https://launchpad.net/~stefanberger/+archive/ubuntu/swtpm-jammy/+sourcefiles/libtpms/0.9.3-1/libtpms_0.9.3-1.tar.xz

It should also resolve the `--fstack-protector` issue I had seen on some platforms, like hppa here when 0.9.2-3 was built: https://buildd.debian.org/status/package.php?p=libtpms

I think we can close this issue now, @frediz ?

**Comment by frediz** [comment]
Hi @stefanberger ,
sure we can. The Debian maintainer pulled the patch and will soon release 0.9.3 in Debian from what I saw.
Thanks a lot for helping and being quick!
F.
